Stratis.Bitcoin/BlockPulling/LookaheadBlockPuller.cs

```diff
@@ -393,13 +388,84 @@ private void AskBlocks()
+                this.ProcessQueue();
             }
 
             this.logger.LogTrace("(-)");
         }
 
         /// <summary>
+        /// Adds requests to download blocks to the queue.
+        /// </summary>
+        /// <param name="downloadRequests">
+        /// Array of block descriptions that need to be downloaded. Must not be empty.
+        /// Blocks in the array have to be unique - it is not supported for a single block to be included twice in this array.
+        /// </param>
+        private void QueueRequests(ChainedBlock[] downloadRequests)
+        {
+            this.logger.LogTrace($"({nameof(downloadRequests)}.{nameof(downloadRequests.Length)}:{downloadRequests.Length})");
+
+            lock (this.bufferLock)
+            {
+                for (int i = 0; i < downloadRequests.Length; i++)
+                    this.askBlockQueue.Enqueue(downloadRequests[i]);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Asks for blocks if there is a free space or the next block is waiting.
+        /// </summary>
+        /// <remarks>Note that this method relies on the fact that the puller requests are ordered.</remarks>
+        private void ProcessQueue()
+        {
+            this.logger.LogTrace("()");
+
+            var requestsToAsk = new List<ChainedBlock>();
+            lock (this.bufferLock)
+            {
+                // We estimate the space needed for the next block as the average size of unconsumed blocks 
+                // the puller currently holds. If it holds no blocks, it does not matter, there is certainly 
+                // a space for new block so we use dummy value 0.
+                long avgBlockSize = this.currentBufferedCount != 0 ? this.currentBufferedSize / this.currentBufferedCount : 0;
+
+                // We want to be sure that location does not move while we distribute the requests.
+                lock (this.locationLock)
+                {
+                    while (this.askBlockQueue.Count > 0)
+                    {
+                        ChainedBlock request = this.askBlockQueue.Peek();
+                        bool isNextBlock = request.Height == this.location.Height + 1;
+
+                        // Buffer is full if the current buffered size plus expected size of all blocks we will ask, 
+                        // including the next request we are considering, is greater than the max limit.
+                        bool bufferFull = this.currentBufferedSize + (requestsToAsk.Count + 1) * avgBlockSize > this.MaxBufferedSize;
+
+                        if (isNextBlock || !bufferFull)
+                        {
+                            requestsToAsk.Add(request);
+                            this.askBlockQueue.Dequeue();
+                        }
+                        else
+                        {
+                            // The buffer is either full, so we do not want to ask for more blocks. 
```

Comment:
We do have issues with reorgs in all code, we do have an issue for a memory leak somewhere but I believe there will be more issues. The good thing here is that it seems that reorg will just cause requested blocks from different chain to be downloaded. So they will arrive earlier then they would otherwise (if we cleaned the list). So there seems to be not even a memory leak in this queue in case of reorg. So maybe there is no problem to solve here. But again, I think there are problems with reorgs and many yet to be uncovered.